src/main/kotlin/app/ui/AuthState.kt

```diff
@@ -6,51 +6,80 @@ package app.ui
-        var username = ""
-        var password = ""
-        var isCredentialsUnknown = !Configurator.isValidCredentials()
-
-        if (isCredentialsUnknown) {
-            println("Enter Sourcerer account username:")
-            username = readLine() ?: ""
-            Configurator.setUsernameCurrent(username)
+        if (!Configurator.isValidCredentials()) {
+            getUsername()
+            getPassword()
         }
 
-        while (true) {
-            if (isCredentialsUnknown) {
-                println("Enter Sourcerer account password:")
-                password = PasswordHelper.readPassword()
-                Configurator.setPasswordCurrent(password)
-            }
+        while (!tryAuth() && !connectionError) {
+            getPassword()
+        }
+    }
+
+    override fun next() {
+        if (!connectionError) {
+            context.changeState(ListRepoState(context))
+        } else {
+            context.changeState(CloseState(context))
+        }
+    }
 
-            println("Authorising...")
-            val result = SourcererApi.getUserBlocking().profileUrl
+    fun getUsername() {
+        println("Enter Sourcerer account username:")
+        username = readLine() ?: ""
+        Configurator.setUsernameCurrent(username)
+    }
 
-            if (result.isNotEmpty()) {  // TODO(anatoly): Check for status 200.
-                println("You are successfully authorised. Your profile page is "
-                        + "http://sourcerer.io/007.")
+    fun getPassword() {
+        println("Enter Sourcerer account password:")
+        password = PasswordHelper.readPassword()
+        Configurator.setPasswordCurrent(password)
+    }
 
-                // Save username and password to config file.
-                if (isCredentialsUnknown) {
-                    Configurator.setUsernamePersistent(username)
-                    Configurator.setPasswordPersistent(password)
-                    Configurator.saveToFile()
-                }
+    fun saveConfigIfChanged() {
+        if (username.isNotEmpty()) {
+            Configurator.setUsernamePersistent(username)
+        }
+        if (password.isNotEmpty()) {
+            Configurator.setPasswordPersistent(password)
+        }
+        if (username.isNotEmpty() || password.isNotEmpty()) {
+            Configurator.saveToFile()
+        }
+    }
 
-                break
+    fun tryAuth(): Boolean {
+        try {
+            doAuth()
+            return true
+        } catch (e: RequestException) {
+            if(e.isAuthError()) {
+                println("Incorrect username or password. Try again.")
+            } else {
+                connectionError = true
+                println("Connection problems. Try again later.")
             }
-
-            isCredentialsUnknown = true
         }
+        return false
     }
 
-    override fun next() {
-        context.changeState(ListRepoState(context))
+    fun doAuth() {
+        println("Authorising...")
```

Comment:
technically doAuth is just a body of a try block inside tryAuth function, I'd say it makes sense to merge these two functions